`rememberCursorPositionProvider` - get rid of the dependency on AWT

Compose 1.3

When we use `Popup`, `TooltipPlacement.CursorPoint` or `rememberCursorPositionProvider`, and use it in an environment without AWT (tests, LWJGL), it should work as with the default configration with AWT:

```
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.TooltipArea
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Button
import androidx.compose.material.Surface
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.*
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.singleWindowApplication

@OptIn(ExperimentalFoundationApi::class)
fun main() = singleWindowApplication {
    TooltipArea(
        tooltip = {
            // composable tooltip content
            Surface(
                modifier = Modifier.shadow(4.dp),
                color = Color(255, 255, 210),
                shape = RoundedCornerShape(4.dp)
            ) {
                Text(
                    text = "Tooltip",
                    modifier = Modifier.padding(10.dp)
                )
            }
        }
    ) {
        Button({}){}
    }
}
```

Now, it isn't supported if we run it in tests, and we can encounter a crash:
```
java.lang.IllegalStateException: CompositionLocal LayerContainer not provided
	at ???(Coroutine boundary.?(?)
	at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2.invokeSuspend(Recomposer.kt:503)
	at androidx.compose.runtime.Recomposer$recompositionRunner$2$2.invokeSuspend(Recomposer.kt:883)
	at androidx.compose.runtime.Recomposer$recompositionRunner$2.invokeSuspend(Recomposer.kt:882)
	at androidx.compose.ui.ComposeScene$6.invokeSuspend(ComposeScene.skiko.kt:242)
Caused by: java.lang.IllegalStateException: CompositionLocal LayerContainer not provided
	at androidx.compose.ui.awt.LocalLayerContainer_desktopKt$LocalLayerContainer$1.invoke(LocalLayerContainer.desktop.kt:23)
	at androidx.compose.ui.awt.LocalLayerContainer_desktopKt$LocalLayerContainer$1.invoke(LocalLayerContainer.desktop.kt:22)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at androidx.compose.runtime.LazyValueHolder.getCurrent(ValueHolders.kt:29)
	at androidx.compose.runtime.LazyValueHolder.getValue(ValueHolders.kt:31)
	at androidx.compose.runtime.ComposerImpl.resolveCompositionLocal(Composer.kt:1981)
	at androidx.compose.runtime.ComposerImpl.consume(Composer.kt:1949)
	at androidx.compose.ui.window.DesktopPopup_desktopKt.rememberCursorPositionProvider-B5uucgQ(DesktopPopup.desktop.kt:136)
	at androidx.compose.foundation.TooltipPlacement$CursorPoint.positionProvider(TooltipArea.desktop.kt:225)
	at androidx.compose.foundation.TooltipArea_desktopKt.TooltipArea(TooltipArea.desktop.kt:150)
	at androidx.compose.foundation.TooltipArea_desktopKt$TooltipArea$7.invoke(TooltipArea.desktop.kt)
	at androidx.compose.foundation.TooltipArea_desktopKt$TooltipArea$7.invoke(TooltipArea.desktop.kt)
	at androidx.compose.runtime.RecomposeScopeImpl.compose(RecomposeScopeImpl.kt:145)
	at androidx.compose.runtime.ComposerImpl.recomposeToGroupEnd(Composer.kt:2375)
	at androidx.compose.runtime.ComposerImpl.skipToGroupEnd(Composer.kt:2666)
	at compose.components.MoreIconKt.MoreIcon(MoreIcon.kt:23)
	at compose.components.ToolVersionsGridKt$GridMenu$7.invoke(ToolVersionsGrid.kt:466)
	at compose.components.ToolVersionsGridKt$GridMenu$7.invoke(ToolVersionsGrid.kt:465)
	at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:107)
	at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:34)
	at jewel.DropdownKt$Dropdown$7$1$1.invoke(Dropdown.kt:295)
	at jewel.DropdownKt$Dropdown$7$1$1.invoke(Dropdown.kt:295)
	at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:107)
	at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:34)
	at androidx.compose.runtime.CompositionLocalKt.CompositionLocalProvider(CompositionLocal.kt:228)
	at jewel.DropdownKt.Dropdown-WSk2ftU(Dropdown.kt:293)
	at compose.jewel.MenuKt.Menu-7vwJh_E(Menu.kt:48)
	at compose.components.ToolVersionsGridKt.GridMenu(ToolVersionsGrid.kt:460)
	at compose.components.ToolVersionsGridKt.GridItem-DzVHIIc(ToolVersionsGrid.kt:339)
	at compose.components.ToolVersionsGridKt.access$GridItem-DzVHIIc(ToolVersionsGrid.kt:1)
	at compose.components.ToolVersionsGridKt$GridItem$10.invoke(ToolVersionsGrid.kt)
	at compose.components.ToolVersionsGridKt$GridItem$10.invoke(ToolVersionsGrid.kt)
	at androidx.compose.runtime.RecomposeScopeImpl.compose(RecomposeScopeImpl.kt:145)
	at androidx.compose.runtime.ComposerImpl.recomposeToGroupEnd(Composer.kt:2375)
	at androidx.compose.runtime.ComposerImpl.skipToGroupEnd(Composer.kt:2666)
```

I wasn't able to reproduce it in tests, but the fix is to not use `LocalLayerContainer.current` in `rememberCursorPositionProvider`. 

For that, we either need to move to pure Compose implementation (using `Modifier.pointerInput`), or inject the global cursor provider via `Platform` class.

It is a part of 2 bigger works:
- moving all AWT dependencies from `ComposeScene` (the facade for Compose) into a higher level
- stabilizing `ui-test`

Reported [here](https://jetbrains.slack.com/archives/C01G87BMA2Z/p1677835531500279)

--- a/compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/TooltipAreaTest.kt
+++ b/compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/TooltipAreaTest.kt
@@ -1,0 +1,66 @@
+/*
+ * Copyright 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.foundation
+
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.size
+import androidx.compose.foundation.text.BasicText
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.platform.testTag
+import androidx.compose.ui.test.ExperimentalTestApi
+import androidx.compose.ui.test.junit4.createComposeRule
+import androidx.compose.ui.test.onNodeWithTag
+import androidx.compose.ui.test.performMouseInput
+import androidx.compose.ui.unit.dp
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.runBlocking
+import org.junit.Rule
+import org.junit.Test
+
+@OptIn(ExperimentalFoundationApi::class, ExperimentalTestApi::class)
+internal class TooltipAreaTest {
+    @get:Rule
+    val rule = createComposeRule()
+
+    // https://github.com/JetBrains/compose-jb/issues/2821
+    @Test
+    fun `simple tooltip is shown`(): Unit = runBlocking(Dispatchers.Main) {
+        rule.setContent {
+            TooltipArea(
+                tooltip = {
+                    Box {
+                        BasicText(
+                            text = "Tooltip",
+                            modifier = Modifier.testTag("tooltipText")
+                        )
+                    }
+                }
+            ) {
+                BasicText("Text", modifier = Modifier.size(50.dp).testTag("elementWithTooltip"))
+            }
+        }
+
+
+        rule.onNodeWithTag("elementWithTooltip").performMouseInput {
+            moveTo(Offset(30f, 40f))
+        }
+        rule.waitForIdle()
+
+        rule.onNodeWithTag("tooltipText").assertExists()
+    }
+}

--- a/compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/TooltipAreaTest.kt
+++ b/compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/TooltipAreaTest.kt
@@ -55,7 +55,6 @@ internal class TooltipAreaTest {
             }
         }
 
-
         rule.onNodeWithTag("elementWithTooltip").performMouseInput {
             moveTo(Offset(30f, 40f))
         }

--- a/compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/DesktopPopup.desktop.kt
+++ b/compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/DesktopPopup.desktop.kt
@@ -19,7 +19,7 @@ import androidx.compose.runtime.Composable
 import androidx.compose.runtime.remember
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.ExperimentalComposeUiApi
-import androidx.compose.ui.awt.LocalLayerContainer
+import androidx.compose.ui.LocalComposeScene
 import androidx.compose.ui.geometry.Offset
 import androidx.compose.ui.platform.LocalDensity
 import androidx.compose.ui.unit.Dp
@@ -30,24 +30,15 @@ import androidx.compose.ui.unit.IntSize
 import androidx.compose.ui.unit.LayoutDirection
 import androidx.compose.ui.unit.dp
 import androidx.compose.ui.unit.round
-import java.awt.MouseInfo
-import javax.swing.SwingUtilities.convertPointFromScreen
 import kotlin.math.roundToInt
 
 /**
- * Returns a remembered value of the mouse cursor position.
+ * Returns a remembered value of the mouse cursor position or null if cursor is not inside a scene.
  */
 @Composable
-private fun rememberCursorPosition(): Offset {
-    val component = LocalLayerContainer.current
-    return remember {
-        val awtMousePosition = MouseInfo.getPointerInfo().location
-        convertPointFromScreen(awtMousePosition, component)
-        Offset(
-            (awtMousePosition.x * component.density.density),
-            (awtMousePosition.y * component.density.density)
-        )
-    }
+private fun rememberCursorPosition(): Offset? {
+    val scene = LocalComposeScene.current
+    return remember { scene.pointerPositions.values.firstOrNull() }
 }
 
 /**
@@ -68,7 +59,15 @@ fun rememberCursorPositionProvider(
     val windowMarginPx = windowMargin.roundToPx()
     val cursorPosition = rememberCursorPosition()
 
-    return remember(cursorPosition, offsetPx, alignment, windowMarginPx){
+    if (cursorPosition == null) {
+        // if cursor is outside the scene, show popup under the parent component
+        return rememberComponentRectPositionProvider(
+            alignment = alignment,
+            offset = offset
+        )
+    }
+
+    return remember(cursorPosition, offsetPx, alignment, windowMarginPx) {
         PopupPositionProviderAtPosition(
             positionPx = cursorPosition,
             isRelativeToAnchor = false,

--- a/compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/DesktopCursorPositionTest.kt
+++ b/compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/DesktopCursorPositionTest.kt
@@ -78,7 +78,6 @@ internal class DesktopCursorPositionTest {
                 }
             }
         }
-        rule.waitForIdle()
 
         rule.onNodeWithTag("testBox").performMouseInput {
             moveTo(Offset(30f, 40f))

--- a/compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/DesktopCursorPositionTest.kt
+++ b/compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/DesktopCursorPositionTest.kt
@@ -1,0 +1,89 @@
+/*
+ * Copyright 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.window
+
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.size
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.input.pointer.PointerEventType
+import androidx.compose.ui.input.pointer.onPointerEvent
+import androidx.compose.ui.platform.testTag
+import androidx.compose.ui.test.ExperimentalTestApi
+import androidx.compose.ui.test.junit4.createComposeRule
+import androidx.compose.ui.test.onNodeWithTag
+import androidx.compose.ui.test.performMouseInput
+import androidx.compose.ui.unit.IntOffset
+import androidx.compose.ui.unit.IntRect
+import androidx.compose.ui.unit.IntSize
+import androidx.compose.ui.unit.LayoutDirection
+import androidx.compose.ui.unit.dp
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.runBlocking
+import org.junit.Rule
+import org.junit.Test
+
+@OptIn(ExperimentalComposeUiApi::class, ExperimentalTestApi::class)
+internal class DesktopCursorPositionTest {
+    @get:Rule
+    val rule = createComposeRule()
+
+    private val windowSize = IntSize(200, 200)
+    private val anchorPosition = IntOffset(0, 0)
+    private val anchorSize = IntSize(100, 100)
+    private val popupSize = IntSize(20, 20)
+
+    // https://github.com/JetBrains/compose-jb/issues/2821
+    @Test
+    fun `pointer position with single component`(): Unit = runBlocking(Dispatchers.Main) {
+        var pointerPosition: IntOffset? = null
+
+        rule.setContent {
+            var savePointerPosition by remember { mutableStateOf(false) }
+            Box(
+                modifier = Modifier
+                    .size(200.dp, 200.dp)
+                    .testTag("testBox")
+                    .onPointerEvent(PointerEventType.Enter, onEvent = {
+                        savePointerPosition = true
+                    })
+            ) {
+                if (savePointerPosition) {
+                    pointerPosition = rememberCursorPositionProvider().calculatePosition(
+                        IntRect(anchorPosition, anchorSize),
+                        windowSize,
+                        LayoutDirection.Ltr,
+                        popupSize
+                    )
+                }
+            }
+        }
+        rule.waitForIdle()
+
+        rule.onNodeWithTag("testBox").performMouseInput {
+            moveTo(Offset(30f, 40f))
+        }
+        rule.waitForIdle()
+        assertThat(pointerPosition).isEqualTo(IntOffset(30, 40))
+    }
+}

--- a/compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
+++ b/compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021 The Android Open Source Project
+ * Copyright 2023 The Android Open Source Project
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -61,7 +61,6 @@ import androidx.compose.ui.unit.dp
 import androidx.compose.ui.unit.toIntRect
 import kotlin.coroutines.CoroutineContext
 import kotlin.jvm.Volatile
-import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.CoroutineStart
 import kotlinx.coroutines.Dispatchers
@@ -232,6 +231,11 @@ class ComposeScene internal constructor(
         }
 
     private var isClosed = false
+
+    /**
+     * Provides mapping pointer id -> current pointer position inside [ComposeScene]
+     */
+    internal val pointerPositions = mutableMapOf<PointerId, Offset>()
 
     init {
         GlobalSnapshotManager.ensureStarted()
@@ -477,6 +481,30 @@ class ComposeScene internal constructor(
         forEachOwner { it.measureAndLayout() }
         pointerPositionUpdater.beforeEvent(event)
         processPointerInput(event)
+        updatePointerPositions(event)
+    }
+
+    private fun updatePointerPositions(event: PointerInputEvent) {
+        // update positions for pointers that are down + mouse (if it is not Exit event)
+        for (pointer in event.pointers) {
+            if ((pointer.type == PointerType.Mouse && event.eventType != PointerEventType.Exit) ||
+                pointer.down
+            ) {
+                pointerPositions[pointer.id] = pointer.position
+            }
+        }
+        // touches/styluses positions should be removed from [pointerPositions] if they are not down anymore
+        // also, mouse exited ComposeScene should be removed
+        val iterator = pointerPositions.iterator()
+        while (iterator.hasNext()) {
+            val pointerId = iterator.next().key
+            val pointer = event.pointers.find { it.id == pointerId } ?: continue
+            if ((pointer.type != PointerType.Mouse && !pointer.down) ||
+                (pointer.type == PointerType.Mouse && event.eventType == PointerEventType.Exit)
+            ) {
+                iterator.remove()
+            }
+        }
     }
 
     @Suppress("DEPRECATION")

